gitoid/src/finder.rs

```diff
@@ -0,0 +1,316 @@
+use crate::Error;
+use crate::GitOid;
+use crate::HashAlgorithm;
+use crate::ObjectType;
+use crate::Result;
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::BufReader;
+use std::io::Read;
+use std::io::Seek;
+use std::path::Path;
+use std::path::PathBuf;
+
+/// Struct for finding artifacts based on matching GitOids.
+///
+/// It's constructed with a default hash algorithm and object type. This is so,
+/// if the `Finder` needs to resolve something like a `&Path` or a seekable
+/// reader to a `GitOid`, it knows what settings to apply to the `GitOid`
+/// construction.
+///
+/// The `Finder` works by supplying an iterator of `GitOid`s on construction
+/// against which the finding will be performed.
+///
+/// When finding, you supply an iterator of "providers", which can be:
+///
+/// 1. `GitOid`
+/// 2. `&GitOid`
+/// 3. `(I, GitOid)`
+/// 4. `(I, &GitOid)`
+/// 5. `&Path`
+/// 6. `(&Path, HashAlgorithm)`
+/// 7. `(&Path, ObjectType)`
+/// 8. `(&Path, HashAlgorithm, ObjectType)`
+/// 9. `(I, R) where R: Read + Seek`
+/// 10. `(I, R, HashAlgorithm) where R: Read + Seek`
+/// 11. `(I, R, ObjectType) where R: Read + Seek`
+/// 12. `(I, R, HashAlgorithm, ObjectType) where R: Read + Seek`
+///
+/// With just `GitOid`s (options 1 and 2), the identifiers are the `GitOid`s
+/// themselves.
+///
+/// When pairing `GitOid`s with an "identifier" type (options 3 and 4), the
+/// identifiers indicate what artifact each `GitOid` represents (like a string
+/// name or a path on the filesystem).
+///
+/// When providing a path, optionally with per-path overrides for
+/// `HashAlgorithm` and/or `ObjectType` (options 6, 7, and 8), the paths are
+/// the identifiers, and are also used to generate the `GitOid`s.
+///
+/// When pairing an identifier type with a reader and optional overrides for
+/// `HashAlgorithm` and/or `ObjectType` (options 9, 10, 11, and 12), the
+/// identifiers are used to identify each `GitOid`, and the readers are used
+/// to generate the `GitOid`s.
+pub struct Finder {
+    default_hash_algorithm: HashAlgorithm,
+    default_object_type: ObjectType,
+    gitoids: HashSet<GitOid>,
+}
+
+impl Finder {
+    /// Construct a new `Finder` for a set of `GitOid`s.
+    pub fn for_gitoids<'a, I>(
+        default_hash_algorithm: HashAlgorithm,
+        default_object_type: ObjectType,
+        gitoids: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = &'a GitOid>,
+    {
+        Finder {
+            default_hash_algorithm,
+            default_object_type,
+            gitoids: gitoids.into_iter().copied().collect(),
```

Comment:
One discussion we could have is the fact that the finder copies the `GitOid`s it's matching against. `GitOid`s are pretty small, so the idea is that this is cheap, but we may consider if there are ways to refactor to avoid the copying.

Reply:
At this scale I'm inclined to take the hit on the small copy.  Don't get me wrong, there are times and places to sweat every memcpy ... but since we doing a *lot* of syscalls to 'find' I suspect the copy here is 'comparatively cheap.

Reply:
We could certainly leave it as-is for now and change it later if we measure it and the copies are a problem.